### PR TITLE
feat(sampling): add HY3 fused sampling for SM100

### PR DESCRIPTION
**Pull request by piaoyanglink15**
## Summary

This PR adds `fused_sampling_from_logits_hy3`, an opt-in HY3 sampling API with an SM100/B200-specialized CUDA path and a portable fallback. Existing FlashInfer sampling entry points are unchanged.

The operator fuses repetition penalty, temperature, optional softmax, top-k/top-p filtering, random sampling, and packed history-mask writeback. It provides a temperature-only fast path, a two-stage full path, caller-provided workspace/output buffers for graph-stable addresses, JIT/AOT registration, trace support, tests, and a standalone benchmark against current FlashInfer paths.

The CUDA implementation is derived from Tencent/HPC-Ops commit `1cd332980ed46bd0172091c1c35d55338fcae47a`; the derived header retains the original copyright, MIT SPDX identifier, and source revision.

This is the sampling-only half of the previously combined submission. The HY3 QK RMSNorm/RoPE KV-store operator is submitted separately.

## Duplicate-work check

I searched open FlashInfer PRs for fused logits sampling, top-k/top-p sampling, and repetition penalty. PR #1561 covers radix top-k as an individual stage, but I did not find an open PR implementing this packed-history HY3 end-to-end contract.

## Tests

Run from the repository root on an SM100/B200 CUDA GPU:

```bash
python -m pytest -q tests/utils/test_sampling.py -k fused_sampling_from_logits_hy3
python -m pytest -q \
  tests/trace/test_fi_trace_template_consistency.py \
  tests/trace/test_template_init.py \
  tests/trace/test_template_registry.py \
  -k fused_sampling_hy3
```

Results:

```text
25 passed, 1911 deselected
6 passed, 1 skipped, 1325 deselected
```

Ruff 0.12.8 lint and format checks pass for all changed Python files.

## Benchmark

Environment: NVIDIA B200, BF16 logits, vocabulary size 120832, CUDA-event timing after the preferred CUPTI timer reported unavailable, cold L2, 20 warmup iterations, and 100 samples. Setup, allocation, mask construction/packing, and JIT compilation are outside timing. Both providers use internal RNG.

```bash
python -m benchmarks.bench_sampling_hy3 \
  --batches 1 8 32 128 512 \
  --warmup 20 --iters 100
```

Temperature compares against the faster of FlashInfer's `softmax + sampling_from_probs` and temperature scaling plus `sampling_from_logits`. Full sampling compares against FP32 repetition penalty/temperature followed by FlashInfer's current top-k-first small-K path and equivalent history writeback.

| batch | HY3 temperature (us) | FlashInfer best (us) | speedup | HY3 full (us) | FlashInfer full (us) | speedup |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 12.304 | 52.320 | 4.252x | 39.904 | 153.712 | 3.852x |
| 8 | 15.328 | 53.312 | 3.478x | 54.176 | 170.832 | 3.153x |
| 32 | 24.640 | 67.440 | 2.737x | 135.920 | 179.232 | 1.319x |
| 128 | 53.472 | 111.776 | 2.090x | 424.848 | 335.520 | 0.790x |
| 512 | 151.760 | 441.376 | 2.908x | 1621.040 | 1128.848 | 0.696x |

The full path is optimized for low-to-moderate decode batches; the benchmark intentionally reports the crossover where the existing FlashInfer composition is faster.

## AI assistance disclosure

AI assistance was used to help prepare and validate this change. I reviewed the changed lines and test results and am responsible for the contribution.

### CODE REVIEW

**Comment by piaoyanglink15** [comment]
Hi maintainers, the required PR Test is currently blocked by the repository's CI authorization gate for the current commit. Local NVIDIA B200 operator tests pass (74 passed), trace consistency and registry tests pass (1145 passed, 194 skipped), and all pre-commit hooks pass. Could someone with CI permission please start the full test run? Thanks!

**Comment by piaoyanglink15** [comment]
@qsang-nv @bkryu @saltyminty Could you please take a look at this PR when you have time? Feedback on the FlashInfer integration/API placement and the B200-specific kernel implementation would be greatly appreciated. Thank you!

**Comment by piaoyanglink15** [comment]
This PR was automatically closed when the cross-repository head branch was renamed to the sampling-only name. The unchanged sampling submission continues in #4891.
